added openPMD-api support for I/O

guj opened the pull request:
## Summary
The goal of this pull request is to enable I/O support through the openPMD-api  library

# 
The proposed changes:
- [x] add new capabilities to AMReX

guj committed:
added openPMD-api support for I/O

WeiqunZhang commented on Src/Base/AMReX_PlotFileUtil.H:
```diff
@@ -16,6 +16,10 @@
 #include <string>
 #include <memory>
 
+#ifdef AMREX_USE_OPENPMD_API
```
Move this header above the C++'s headers.

guj replied:
done.

WeiqunZhang commented on Src/Extern/openPMD-api/AMReX_ParticlesOPENPMD.H:
```diff
@@ -0,0 +1,80 @@
+};
+
+
+void CountParticles()
```
We need to put his inside namespace amrex.

guj replied:
this file is included in "AMReX_ParticleContainer.H",  in which the namespace amrex was declared.

WeiqunZhang replied:
Oh I see. Maybe a comment can be added at the top of this file.

guj replied:
done.

WeiqunZhang commented on Src/Extern/openPMD-api/AMReX_PlotFileOPENPMD.cpp:
```diff
@@ -0,0 +1,593 @@
+
+#include <openPMD/openPMD.hpp>
+#include <AMReX_PlotFileUtilOPENPMD.H>
+#include <regex>
```
Let's reorder the headers in a way that amrex headers are followed by openPMD headers and then C++ headers. Please also use this ordering in other places.

guj replied:
done.

WeiqunZhang commented on Src/Extern/openPMD-api/AMReX_ParticlesOPENPMD.H:
```diff
@@ -0,0 +1,80 @@
+    }
+}
+
+AMReX_PtlCounter m_PtlCounter;
```
This would result in multiple definitions if this header is included in more than one translation units.

guj replied:
This eventually is a member variable of class ParticleContainer_impl, as this file is included in "AMReX_ParticleContainer.H".

ax3l replied:
Instead of plain text including this into another class, which will complicate our transition to C++ modules and is a bit surprising in general, consider making this a C++ mixin class:
  https://en.wikipedia.org/wiki/Mixin

Simply derive `ParticleContainer_impl` from this.

guj replied:
The extra code in this file is a function that figures out ptl offsets  for each processors.  I think using mixin possibly will be an overkill,  and clients need to change their code when initialing ParticleContainer_impl related classes. 

Since this is only used by openPMD-api,  I will move this function into the  amex::openpmd_api scope.  Leave the AMReX_ParticleContainer.H as is.

guj committed:
make AMREX_USE_OPENPMD_API flag visible

guj committed:
ordered include file orders (amrex then openpmd then cpp)

guj committed:
fixed spelling errors

guj committed:
Added a comment to address namespace

guj committed:
attempt to remove trailing whitespace

WeiqunZhang commented on Src/Extern/openPMD-api/AMReX_ParticlesOPENPMD.H:
```diff
@@ -0,0 +1,85 @@
+
+  std::vector<unsigned long long> m_ParticleCounterByLevel;
+
+  unsigned long GetTotalNumParticles () const { return m_Total;}
```
```suggestion
  [[nodiscard]] unsigned long long GetTotalNumParticles () const { return m_Total;}
```

WeiqunZhang committed:
Fix warnings

WeiqunZhang commented on Src/Extern/openPMD-api/AMReX_ParticlesOPENPMD.H:
```diff
@@ -0,0 +1,85 @@
+
+  for (auto currentLevel = 0; currentLevel <= this->finestLevel(); currentLevel++)
+    {
+      long numParticles = 0; // numParticles in this processor
```
I wonder if we should use `long long` because of the size of `long` on Windows. And I think the following loop can be replaced by
```
auto numParticles = static_cast<long long>(this->NumberOfParticlesAtLevel(currentLevel, false, false));
```

WeiqunZhang commented on Src/Extern/openPMD-api/AMReX_ParticlesOPENPMD.H:
```diff
@@ -0,0 +1,85 @@
+
+    sum = 0;
+    offset = 0;
+    for (int i=0;  i<result.size();  i++) {
```
```suggestion
    for (std::size_t i=0;  i<result.size();  i++) {
```

WeiqunZhang commented on Src/Extern/openPMD-api/AMReX_PlotFileUtilOPENPMD.H:
```diff
@@ -0,0 +1,425 @@
+    ////////////////////////////
+    struct AMReX_VarNameParser
+    {
+      AMReX_VarNameParser(std::string varname);
```
```suggestion
      AMReX_VarNameParser(std::string varname);
      virtual ~AMReX_VarNameParser () = default;
```

WeiqunZhang reviewed (changes requested):
I actually have more changes to fix various compiler warnings. Do you want me to post as comments or push to your branch?

WeiqunZhang committed:
Add CI test

WeiqunZhang committed:
Merge branch 'development' into use_openpmd-api

WeiqunZhang committed:
Add GNU Make support

guj commented:
> I actually have more changes to fix various compiler warnings. Do you want me to post as comments or push to your branch?

Just saw you pushed. Thanks.

guj committed:
bug fix

guj committed:
Added function SetWriter

WeiqunZhang committed:
Merge branch 'development' into use_openpmd-api

WeiqunZhang committed:
Merge branch 'use_openpmd-api' of github.com:guj/amrex into use_openpmd-api

WeiqunZhang committed:
Add a missing file to CMakeLists.txt

WeiqunZhang committed:
update CI

WeiqunZhang commented on Src/Extern/openPMD-api/AMReX_PlotFileUtilOPENPMD.H:
```diff
@@ -0,0 +1,433 @@
+    namespace helper
+    {
+      inline void
+      setOpenPMDUnit ( openPMD::Mesh mesh, std::string const& field_name )
```
This might be too specific for WarpX.

guj replied:
OK I. can remove this and let apps handle it.

WeiqunZhang commented on Src/Extern/openPMD-api/AMReX_PlotFileUtilOPENPMD.H:
```diff
@@ -0,0 +1,433 @@
+
+        // temporary resolution
+#if AMREX_SPACEDIM == 1
+        vs const axisLabels{"z"};
```
I understand this is a virtual function. But I am wondering if it would be less confusing to simply use `AMREX_D_DECL("x","y","z")` than the WarpX convention.

guj replied:
Sure, I can use AMREX_D_DECL("x","y","z") as default.

WeiqunZhang commented on Src/Extern/openPMD-api/AMReX_PlotFileUtilOPENPMD_PTLImpl.H:
```diff
@@ -0,0 +1,661 @@
+    {
+      BL_PROFILE("amrex::openpmd_api::WriteParticles()");
+
+      amrex::Print()<<"   Real: "<<pc.NStructReal<<" "<<pc.NumRealComps()<<"\n";
```
We need to follow the xSDK policy that the user must be able to disable print statements. So could we put them inside `if (amrex::Verbose())`?

guj replied:
I will remove these. Was for debugging.

guj committed:
removed setOpenPMDUnit, and it will be handled by applications
set default axis labels to be "x/y/z" for 1/2/3-Dims
other behaviours should be done in application codes
removed print statements

ax3l commented on .github/workflows/clang.yml:
```diff
@@ -166,6 +166,77 @@ jobs:
+        cd openPMD-api-0.15.2
+        mkdir build
+        cd build
+        cmake .. -DopenPMD_USE_PYTHON=OFF -DCMAKE_CXX_COMPILER_LAUNCHER=ccache
```
Slightly faster CI:
```suggestion
        cmake ..  \
          -DCMAKE_CXX_COMPILER_LAUNCHER=ccache  \
          -DopenPMD_BUILD_EXAMPLES=OFF  \
          -DopenPMD_BUILD_CLI_TOOLS=OFF  \
          -DopenPMD_BUILD_TESTING=OFF  \
          -DopenPMD_USE_PYTHON=OFF  \
```

ax3l commented on Src/Extern/openPMD-api/AMReX_ParticlesOPENPMD.H:
```diff
@@ -0,0 +1,79 @@
+#ifndef AMREX_PTL_OPENPMD_API_H
+#define AMREX_PTL_OPENPMD_API_H
+
+/**
+*   This file is included from AMReX_ParticleCounter.H
```
Otherwise this will be assigned to `struct AMReX_PtlCounter` below:
```suggestion
/** @file This file is included from AMReX_ParticleCounter.H
```

ax3l commented on Src/Extern/openPMD-api/AMReX_PlotFileUtilOPENPMD_PTLImpl.H:
```diff
@@ -0,0 +1,658 @@
+#ifndef AMREX_PLOTFILE_OPENPMD_PTLS_API_H
+#define AMREX_PLOTFILE_OPENPMD_PTLS_API_H
+
+//extern  std::unique_ptr< amrex::openpmd_api::AMReX_openPMDHandler > amrex::openpmd_api::m_OpenPMDHandler;
```
Unused code?
```suggestion
```

ax3l commented on Tools/GNUMake/packages/Make.openpmd:
```diff
@@ -0,0 +1,20 @@
+CPPFLAGS += -DAMREX_USE_OPENPMD_API
```
Let's simplify to:
```suggestion
CPPFLAGS += -DAMREX_USE_OPENPMD
```

ax3l commented on Tools/CMake/AMReX_Config_ND.H.in:
```diff
@@ -58,6 +58,7 @@
 #cmakedefine AMREX_USE_HDF5
 #cmakedefine AMREX_USE_HDF5_ASYNC
 #cmakedefine AMREX_USE_HDF5_ZFP
+#cmakedefine AMREX_USE_OPENPMD_API
```
Let's simplify to:
```suggestion
#cmakedefine AMREX_USE_OPENPMD
```

ax3l commented on Tools/CMake/AMReXThirdPartyLibraries.cmake:
```diff
@@ -1,4 +1,17 @@
 #
+# openPMD_api related targets
+#
+if (AMReX_OPENPMD_API)
```
Let's simplify to:
```suggestion
if (AMReX_OPENPMD)
```

ax3l commented on Tools/CMake/AMReXThirdPartyLibraries.cmake:
```diff
@@ -1,4 +1,17 @@
 #
+# openPMD_api related targets
```
```suggestion
# openPMD-api related targets
```

ax3l commented on Tools/CMake/AMReXSetDefines.cmake:
```diff
@@ -167,6 +167,11 @@ add_amrex_define(AMREX_USE_HDF5 NO_LEGACY IF AMReX_HDF5)
+#
+# openPMD-api
+#
+add_amrex_define(AMREX_USE_OPENPMD_API NO_LEGACY IF AMReX_OPENPMD_API)
```
Let's simplify to:
```suggestion
add_amrex_define(AMREX_USE_OPENPMD NO_LEGACY IF AMReX_OPENPMD)
```

ax3l commented on Tools/CMake/AMReXOptions.cmake:
```diff
@@ -352,6 +352,10 @@ cmake_dependent_option(AMReX_HDF5_ZFP "Enable ZFP compression in HDF5-based IO"
 print_option(AMReX_HDF5_ZFP)
 
+# openPMD-api
+option(AMReX_OPENPMD_API "Enable I/O through openPMD-api" OFF)
+print_option(AMReX_OPENPMD_API)
```
Let's simplify to:
```suggestion
option(AMReX_OPENPMD "Enable I/O through openPMD-api" OFF)
print_option(AMReX_OPENPMD)
```

ax3l commented on Tools/CMake/AMReXConfig.cmake.in:
```diff
@@ -213,6 +215,10 @@ if (@AMReX_HYPRE@)
    find_dependency(HYPRE 2.20.0 REQUIRED)
 endif ()
 
+if (@AMReX_OPENPMD_API@)
```
Let's simplify to:
```suggestion
if (@AMReX_OPENPMD@)
```

ax3l commented on Tools/CMake/AMReXConfig.cmake.in:
```diff
@@ -136,6 +137,7 @@ set(AMReX_HYPRE                     @AMReX_HYPRE@)
 set(AMReX_PETSC                     @AMReX_PETSC@)
 set(AMReX_HDF5                      @AMReX_HDF5@)
 set(AMReX_HDF5_ZFP                  @AMReX_HDF5_ZFP@)
+set(AMReX_OPENPMD_API               @AMReX_OPENPMD_API@)
```
Let's simplify to:
```suggestion
set(AMReX_OPENPMD                   @AMReX_OPENPMD@)
```

ax3l commented on Tools/CMake/AMReXConfig.cmake.in:
```diff
@@ -85,6 +85,7 @@ set(AMReX_PETSC_FOUND               @AMReX_PETSC@)
 set(AMReX_SUNDIALS_FOUND            @AMReX_SUNDIALS@)
 set(AMReX_HDF5_FOUND                @AMReX_HDF5@)
 set(AMReX_HDF5_ZFP_FOUND            @AMReX_HDF5_ZFP@)
+set(AMReX_OPENPMD_API_FOUND         @AMReX_OPENPMD_API@)
```
Let's simplify to:
```suggestion
set(AMReX_OPENPMD_FOUND             @AMReX_OPENPMD@)
```

ax3l commented on Tests/openPMDTests/ptls/warpxWriter.H:
```diff
@@ -0,0 +1,327 @@
+  mesh_comp.setPosition( relative_cell_pos );
+}
+
+void AMReX_warpxWriter::SetupFields (openPMD::Container< openPMD::Mesh >& meshes,
```
This belongs in WarpX, not here.

ax3l commented on Tests/openPMDTests/ptls/warpxBTD.H:
```diff
@@ -0,0 +1,259 @@
+                                           openPMD::IterationEncoding ie,
+                                           std::string filetype,
+                                           std::string openpmdOptions)
+//  :AMReX_openPMDWriter("{}", ie, filetype, openpmdOptions)
```
```suggestion
```

ax3l commented on Tests/CMakeLists.txt:
```diff
@@ -135,6 +135,10 @@ else()
       list(APPEND AMREX_TESTS_SUBDIRS HDF5Benchmark)
    endif ()
 
+   if (AMReX_OPENPMD_API)
```
```suggestion
   if (AMReX_OPENPMD)
```

ax3l commented on Src/Extern/openPMD-api/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+add_amrex_define(AMREX_USE_OPENPMD_API NO_LEGACY IF AMReX_OPENPMD_API)
```
```suggestion
add_amrex_define(AMREX_USE_OPENPMD NO_LEGACY IF AMReX_OPENPMD)
```

ax3l commented on Src/CMakeLists.txt:
```diff
@@ -200,6 +200,10 @@ if (AMReX_HDF5)
    add_subdirectory(Extern/HDF5)
 endif ()
 
+if (AMReX_OPENPMD_API)
```
```suggestion
if (AMReX_OPENPMD)
```

ax3l commented on .github/workflows/clang.yml:
```diff
@@ -166,6 +166,77 @@ jobs:
+        mkdir build
+        cd build
+        cmake ..                                      \
+            -DAMReX_OPENPMD_API=ON                    \
```
```suggestion
            -DAMReX_OPENPMD=ON                        \
```

ax3l commented on Src/Extern/openPMD-api/AMReX_ParticlesOPENPMD.H:
```diff
@@ -0,0 +1,79 @@
+void CountParticles()
+{
+
+  m_PtlCounter.m_MPISize = amrex::ParallelDescriptor::NProcs();
+  m_PtlCounter.m_MPIRank = amrex::ParallelDescriptor::MyProc();
```
AMReX includes for `ParallelDescriptor`, `ParallelGather`, `<vector>` missing in this file.

guj replied:
This file is currently embedded in AMReX_ParticleContainer.H, which has all the included files declared already.

ax3l commented on Src/Particle/AMReX_ParticleContainer.H:
```diff
@@ -1355,6 +1355,10 @@ public:
 #include "AMReX_ParticlesHDF5.H"
 #endif
 
+#ifdef AMREX_USE_OPENPMD_API
+#include "AMReX_ParticlesOPENPMD.H"
+#endif
```
I would prefer if this was used as a mixin class instead of a textual include.

guj committed:
Update .github/workflows/clang.yml

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update Src/Extern/openPMD-api/AMReX_PlotFileUtilOPENPMD_PTLImpl.H

remove commented code

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update Tools/CMake/AMReX_Config_ND.H.in

reduce AMREX_USE_OPENPMD_API to  AMREX_USE_OPENPMD

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update .github/workflows/clang.yml

reduce AMReX_OPENPMD_API to AMReX_OPENPMD

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update Tools/GNUMake/packages/Make.openpmd

AMREX_USE_OPENPMD instead of AMREX_USE_OPENPMD_API

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
Update Src/CMakeLists.txt

Use AMReX_OPENPMD instead of AMReX_OPENPMD_API

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

guj committed:
shorted flags *OPENPMD_API to *OPENPMD

WeiqunZhang committed:
Apply suggestions from code review

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

WeiqunZhang reviewed (approved):
I don't think this PR could break any existing codes. Since this is big PR and adds a new feature, I am okay with merging it now.

WeiqunZhang committed:
Merge branch 'development' into use_openpmd-api

WeiqunZhang committed:
Merge branch 'use_openpmd-api' of github.com:guj/amrex into use_openpmd-api

asalmgren commented:
@WeiqunZhang @ax3l -- is this an issue that should still be open?
